File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.cc
#include<iostream>
#include<queue>
#include<utility>
#include<tuple>

using namespace std;

#define X first
#define Y second

int dx[4] = {0,1,0,-1};
int dy[4] = {1,0,-1,0};

char board[1000][1000];
int dist[1000][1000][2];
// dist[x][y][0] : 벽을 하나도 부수지 않고 (x,y)까지 오는데 걸리는 비용
// dist[x][y][1] : 벽을 하나만 부수고 (x,y)까지 오는데 걸리는 비용, (x,y)가 벽이라서 부수는 경우 포함

int n, m;

int main(){
    cin.tie(0);
    ios::sync_with_stdio(0);
    cin >> n >> m;
    for (int i = 0; i < n; ++i)
       for (int j = 0; j < m; ++j)
          cin >> board[i][j];
    
  
    for (int i = 0; i < n; ++i)
       for (int j = 0; j < m; ++j)
          dist[i][j][0] = dist[i][j][1] = -1;
    
    dist[0][0][0]= dist[0][0][1]= 0;
    
    queue< tuple<int,int,int> > q; 
    
    q.push({0,0,0});
    
    int tmp=0; 
    
    while(!q.empty())
    {
        int x, y, broken;
        tie(x, y, broken) = q.front(); q.pop();        
        
        
        for(int dir=0; dir<4; dir++)
        {
            int nx = x + dx[dir];
            int ny = y + dy[dir];
            
            if( nx>=n || ny>=m || nx <0|| ny<0) continue;
            if(dist[nx][ny][broken] ==-1 && board[nx][ny] =='0' ) {
            
                dist[nx][ny][broken] = dist[x][y][broken] + 1;
                q.push({nx,ny,broken});
            }
            
            if (!broken && board[nx][ny] =='1' && dist[nx][ny][1]==-1 ){
                
                dist[nx][ny][1] = dist[x][y][broken] + 1;
                q.push({nx,ny,1});
            }
            
            
        }        
    }
     
    
            
    if(dist[n-1][m-1][0] >=0 || dist[n-1][m-1][1] >=0 ) {
        int mn; 
        
        if(dist[n-1][m-1][0]==-1){
            mn= dist[n-1][m-1][1]; 
        } 
        else if(dist[n-1][m-1][1] ==-1 ){
            mn =dist[n-1][m-1][0];
        }
        else {
            mn= min(dist[n-1][m-1][0],dist[n-1][m-1][1]);
        }
        cout<< mn+1 ;
        return 0;        
    }
    
    cout<< -1;
    
    
}

File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {

    public static class Pair {
        int broken;
        int h;
        int w;

        public Pair(int broken, int h, int w) {
            this.broken = broken;
            this.h = h;
            this.w = w;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer stringTokenizer = new StringTokenizer(br.readLine());
        int height = Integer.parseInt(stringTokenizer.nextToken());
        int width = Integer.parseInt(stringTokenizer.nextToken());

        int[][] board = new int[height][width];
        int[][][] dist = new int[2][height][width];

        for (int i = 0; i < height; i++) {
            String line = br.readLine();
            for (int j = 0; j < width; j++) {
                board[i][j] = line.charAt(j) - '0';  // char를 int로 변환
                dist[0][i][j] = dist[1][i][j] = -1;
            }
        }

        int[] dh = {1, 0, -1, 0};
        int[] dw = {0, 1, 0, -1};

        Queue<Pair> q = new LinkedList<>();
        q.offer(new Pair(0,0, 0));
        dist[0][0][0] = dist[1][0][0] = 1; // 0=> 벽 안부숨 , 1 => 벽 부숨


        while (!q.isEmpty()) {
            Pair cur = q.poll();

            for (int dir = 0; dir < 4; dir++) {
                int nh = cur.h + dh[dir];
                int nw = cur.w + dw[dir];

                if (nh >= height || nh < 0 || nw >= width || nw < 0) continue;
                if (board[nh][nw] == 0 && dist[cur.broken][nh][nw] == -1){
                    dist[cur.broken][nh][nw] = dist[cur.broken][cur.h][cur.w] + 1;
                    q.offer(new Pair(cur.broken, nh, nw));
                }

                if(cur.broken == 0 && board[nh][nw] == 1 && dist[cur.broken][nh][nw] == -1 ){
                    dist[1][nh][nw] = dist[cur.broken][cur.h][cur.w] + 1;
                    q.offer(new Pair(1,nh, nw));
                }
            }
        }

        if(dist[0][height-1][width-1] == -1 && dist[1][height-1][width-1] == -1){
            System.out.println(-1);
            return;
        }

        
        if(dist[0][height - 1][width - 1]!= -1 && dist[1][height - 1][width - 1] == -1){
            System.out.println(dist[0][height - 1][width - 1]);
            
        } else if (dist[0][height - 1][width - 1]== -1 && dist[1][height - 1][width - 1] != -1) {
            System.out.println(dist[1][height - 1][width - 1]);
            
        } else {
            System.out.println(Math.min(dist[0][height - 1][width - 1],dist[1][height - 1][width - 1]));
        }


    }
}